onnxruntime/core/providers/dml/DmlExecutionProvider/src/External/DirectMLHelpers/ApiTraits.h

```diff
@@ -2717,4 +2790,658 @@ inline gsl::czstring ToString(DML_OPERATOR_TYPE value)
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_RANDOM_GENERATOR_TYPE value)
+{
+    switch (value)
+    {
+    case DML_RANDOM_GENERATOR_TYPE_PHILOX_4X32_10: return "DML_RANDOM_GENERATOR_TYPE_PHILOX_4X32_10";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_MULTIHEAD_ATTENTION_MASK_TYPE value)
+{
+    switch (value)
+    {
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_NONE: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_NONE";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_LENGTH: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_LENGTH";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_END_START: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_END_START";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_QUERY_SEQUENCE_LENGTH_START_END: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_QUERY_SEQUENCE_LENGTH_START_END";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_BOOLEAN: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_BOOLEAN";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+
+template <typename T>
+inline T FromString(std::string_view value)
+{
+    // Clang will instantiate this template even if it isn't used,
+    // so this static_assert will always fire and break the build.
+    static_assert(false, "Not implemented for this type");
+}
+
+inline std::unordered_map<std::string_view, DML_TENSOR_DATA_TYPE> DML_TENSOR_DATA_TYPE_MAP = {
+   { "DML_TENSOR_DATA_TYPE_UNKNOWN",  DML_TENSOR_DATA_TYPE_UNKNOWN},
+   { "DML_TENSOR_DATA_TYPE_FLOAT32",  DML_TENSOR_DATA_TYPE_FLOAT32},
+   { "DML_TENSOR_DATA_TYPE_FLOAT16",  DML_TENSOR_DATA_TYPE_FLOAT16},
+   { "DML_TENSOR_DATA_TYPE_UINT32",  DML_TENSOR_DATA_TYPE_UINT32},
+   { "DML_TENSOR_DATA_TYPE_UINT16",  DML_TENSOR_DATA_TYPE_UINT16},
+   { "DML_TENSOR_DATA_TYPE_UINT8",  DML_TENSOR_DATA_TYPE_UINT8},
+   { "DML_TENSOR_DATA_TYPE_INT32",  DML_TENSOR_DATA_TYPE_INT32},
+   { "DML_TENSOR_DATA_TYPE_INT16",  DML_TENSOR_DATA_TYPE_INT16},
+   { "DML_TENSOR_DATA_TYPE_INT8",  DML_TENSOR_DATA_TYPE_INT8},
+   { "DML_TENSOR_DATA_TYPE_FLOAT64",  DML_TENSOR_DATA_TYPE_FLOAT64},
+   { "DML_TENSOR_DATA_TYPE_UINT64",  DML_TENSOR_DATA_TYPE_UINT64},
+   { "DML_TENSOR_DATA_TYPE_INT64",  DML_TENSOR_DATA_TYPE_INT64},
+};
+
+template <>
+inline DML_TENSOR_DATA_TYPE FromString(std::string_view value)
+{
+ return DML_TENSOR_DATA_TYPE_MAP[value];
+}
+
+inline std::unordered_map<std::string_view, DML_TENSOR_TYPE> DML_TENSOR_TYPE_MAP = {
+   { "DML_TENSOR_TYPE_INVALID",  DML_TENSOR_TYPE_INVALID},
+   { "DML_TENSOR_TYPE_BUFFER",  DML_TENSOR_TYPE_BUFFER},
+};
+
+template <>
+inline DML_TENSOR_TYPE FromString(std::string_view value)
+{
+ return DML_TENSOR_TYPE_MAP[value];
+}
+
+inline std::unordered_map<std::string_view, DML_OPERATOR_TYPE> DML_OPERATOR_TYPE_MAP = {
+   { "DML_OPERATOR_INVALID",  DML_OPERATOR_INVALID},
```

Comment:
Indent 4 spaces.
Delete extra space between string and enum.

`    { "DML_OPERATOR_INVALID", DML_OPERATOR_INVALID},`

Reply:
Appears we also typically (not always, but usually) put opening braces on a new line. e.g.:

```
static DML_DEPTH_SPACE_ORDER GetDepthSpaceOrder(std::string_view name)
{
    constexpr StringUtil::WideNameAndIndex mapping[] =
    {
        {L"DEPTH_COLUMN_ROW", DML_DEPTH_SPACE_ORDER_DEPTH_COLUMN_ROW},
        {L"COLUMN_ROW_DEPTH", DML_DEPTH_SPACE_ORDER_COLUMN_ROW_DEPTH},
    };

    return GetEnumFromTestValue<DML_DEPTH_SPACE_ORDER>(name, mapping, "Invalid DML_DEPTH_SPACE_ORDER.");
}

static DML_INTERPOLATION_MODE GetInterpolationMode(std::string_view name)
{
    constexpr StringUtil::WideNameAndIndex mapping[] =
    {
        {L"NN", DML_INTERPOLATION_MODE_NEAREST_NEIGHBOR},
        {L"LINEAR", DML_INTERPOLATION_MODE_LINEAR},
    };

    return GetEnumFromTestValue<DML_INTERPOLATION_MODE>(name, mapping, "Invalid DML_INTERPOLATION_MODE.");
}

static DML_REDUCE_FUNCTION GetReduceFunction(std::string_view name)
{
    constexpr StringUtil::WideNameAndIndex mapping[] =
    {
        {L"ARGMAX", DML_REDUCE_FUNCTION_ARGMAX},
        {L"ARGMIN", DML_REDUCE_FUNCTION_ARGMIN},
        {L"AVERAGE", DML_REDUCE_FUNCTION_AVERAGE},
        {L"L1", DML_REDUCE_FUNCTION_L1},
        {L"L2", DML_REDUCE_FUNCTION_L2},
        {L"LOG_SUM", DML_REDUCE_FUNCTION_LOG_SUM},
        {L"LOG_SUM_EXP", DML_REDUCE_FUNCTION_LOG_SUM_EXP},
        {L"MAX", DML_REDUCE_FUNCTION_MAX},
        {L"MIN", DML_REDUCE_FUNCTION_MIN},
        {L"MULTIPLY", DML_REDUCE_FUNCTION_MULTIPLY},
        {L"SUM", DML_REDUCE_FUNCTION_SUM},
        {L"SUM_SQUARE", DML_REDUCE_FUNCTION_SUM_SQUARE},
    };
```